### PR TITLE
Implement Arbitrary for units types

### LINKED ISSUE #3705: units: Implement `Arbitrary` for all types
If #3682 merges you can use `cotrib/api.sh units types` to get a list of types and you check the api files (`api/units/all-features.txt`) to see which are done.

### PR DESCRIPTION
**Pull request by shinghim**
Implement Arbitrary for the rest of the types in `units`. Also moved the implementation in `FeeRate` right before the `tests` module

Closes #3705

### CODE REVIEW

**Comment by tcharding** [comment]
Would be mad to prove that we have got everything by adding a test to the `units/tests/api.rs` file. Perhaps something like

```rust
#[test]
#[cfg(feature = "arbitrary")]
fn arbitrary() {
    use arbitrary::{Arbitrary, Unstructured};

    impl<'a> Arbitrary<'a> for Types {
        fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
            let a = Types {
                a: Enums::arbitrary(u)?,
                b: Structs::arbitrary(u)?,
            };
            Ok(a)
        }
    }

    impl<'a> Arbitrary<'a> for Structs {
        fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
            let a = Structs {
                a: Amount::arbitary(u)?,
                // Skip the `Display` type.
                b: Amount::MAX.display_in(amount::Denomination::Bitcoin),
                c: SignedAmount::arbitary(u)?,
                d: BlockHeight::arbitary(u)?,
                e: BlockInterval::arbitary(u)?,
                f: FeeRate::arbitary(u)?,
                g: absolute::Height::arbitary(u)?,
                h: absolute::Time::arbitary(u)?,
                i: relative::Height::arbitary(u)?,
                j: relative::Time::arbitary(u)?,
                k: Weight::arbitary(u)?,
            };
            Ok(a)
        }
    }

    impl<'a> Arbitrary<'a> for Enums {
        fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
            let a = Enums {
                a: amount::Denomination::arbitrary(u),
            };
            Ok(a)
        }
    }
}
```

(Not quite working but I think you get the idea.)

**Comment by shinghim** [comment]
@tcharding i like the idea of validating that these impls exist. With a test, every time a new test is introduced, it'd have to be added to that test. What about something that matches the output of `contrib/api.sh units types` and `api/units/all-features.txt` like from the description of #3705?

**Comment by tcharding** [comment]
The point of the `tests/api.rs` file is to do various checks so it will have to be update anyways irrespective of any arbitrary stuff we add. Separately, if you want to come up with some way to test that we tested everything correctly I'm ok with that!

File: units/src/locktime/absolute.rs [hunk 1]
```diff
@@ -384,6 +386,36 @@ impl From<ConversionError> for ParseError {
...
+
+#[cfg(feature = "arbitrary")]
+impl<'a> Arbitrary<'a> for Time {
+    fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        let choice = u.int_in_range(0..=2)?;
+        match choice {
+            0 => Ok(Time::MIN),
+            1 => Ok(Time::MAX),
+            _ => {
+                let t = u.int_in_range(LOCK_TIME_THRESHOLD..=Time::MAX.to_consensus_u32())?;
```
Comment (yancyribbens) [inline units/src/locktime/absolute.rs:412]: This doesn't seem right.  MAX is defied as `LOCK_TIME_THRESHOLD - 1`. so MAX is less than `LOCK_TIME_THRESHOLD` if I'm reading this right.  So you have a range of only two values and the larger value in the range first?

Comment (shinghim) [reply units/src/locktime/absolute.rs:412]: I think you are looking at `Height::MAX`; `Time::MAX` is defined as `Time(u32::MAX)`

File: units/src/locktime/absolute.rs [hunk 2]
```diff
@@ -384,6 +386,36 @@ impl From<ConversionError> for ParseError {
...
 
+#[cfg(feature = "arbitrary")]
+impl<'a> Arbitrary<'a> for Height {
+    fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        let choice = u.int_in_range(0..=2)?;
+        match choice {
+            0 => Ok(Height::MIN),
+            1 => Ok(Height::MAX),
+            _ => {
+                let h = u.int_in_range(0..=Height::MAX.to_consensus_u32())?;
```
[the remark this answers is not in the document]
Comment (shinghim) [reply units/src/locktime/absolute.rs:397]: i don't think so. This range should include all possible values for `Height`. Maybe `Height::MIN.to_consensus_u32()..=Height::MAX.to_consensus_u32()` would be more clear?

**Comment by shinghim** [comment]
@tcharding on a related note to checking that all types have `Arbitrary` implemented, would you by chance know why in `units/all-features.txt`:
```
impl<'a> arbitrary::Arbitrary<'a> for bitcoin_units::Amount
impl<'a> arbitrary::Arbitrary<'a> for bitcoin_units::SignedAmount
impl<'a> arbitrary::Arbitrary<'a> for bitcoin_units::block::BlockHeight
impl<'a> arbitrary::Arbitrary<'a> for bitcoin_units::block::BlockInterval
impl<'a> arbitrary::Arbitrary<'a> for bitcoin_units::fee_rate::FeeRate
impl<'a> arbitrary::Arbitrary<'a> for bitcoin_units::locktime::absolute::Height
impl<'a> arbitrary::Arbitrary<'a> for bitcoin_units::locktime::absolute::Time
impl<'a> arbitrary::Arbitrary<'a> for bitcoin_units::locktime::relative::Height
impl<'a> arbitrary::Arbitrary<'a> for bitcoin_units::locktime::relative::Time
impl<'a> arbitrary::Arbitrary<'a> for bitcoin_units::weight::Weight
```

`Amount` and `SignedAmount` don't have the full namespace, but the other types do? I've spent a bit of time looking into it but can't quite figure out why `cargo public-api` lists them out differently. My plan was to make a script that compares these to the output of `./contrib/api.sh units types_no_err`, but this discrepancy makes it a bit harder to do it cleanly

File: units/src/locktime/absolute.rs (hunk 1 above)
Comment (yancyribbens) [reply units/src/locktime/absolute.rs:412]: Oh ok, then that checks out.  Sorry for the confusion.

File: units/src/locktime/absolute.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (yancyribbens) [reply units/src/locktime/absolute.rs:397]: right, I'd probably do:

```
let min = Height::MIN.to_consensus_u32()
let max = Height::MAX.to_consensus_u32()
let h = u.int_in_range(0..=Height::MAX.to_consensus_u32(min..max))?;
```

[the remark this answers is not in the document]
Comment (yancyribbens) [reply units/src/locktime/absolute.rs:397]: Not that MIN and MAX are ever to likely change, but it's good practice I think.

**Comment by shinghim** [comment]
@tcharding is it cool if i make a new issue for the Arbitrary impl validation? Im trying to think of a way that we could put it into a script similar to `check-for-api-changes.sh`

**Comment by shinghim** [comment]
also AFAICT about the namespace thing, it's just how Rust handles files in different directories

**Comment by tcharding** [comment]
> is it cool if i make a new issue for the Arbitrary impl validation?

Sure. I'm likely to just PR the idea I posted above though because we are super close to 1.0-ing `units` and I'd like to see this `Arbitrary` stuff verified like everything else is in the `api` test file.

**Comment by shinghim** [comment]
@tcharding ah i see. I can add the Arbitrary verification here since im already making related changes and then just come back to seeing if theres a way to do it via script later

File: units/tests/api.rs [hunk 3]
```diff
@@ -263,3 +265,43 @@ fn regression_default() {
...
+        };
+        Ok(a)
+    }
+}
+
+#[cfg(feature = "arbitrary")]
+impl<'a> Arbitrary<'a> for Enums {
+    fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        let a = Enums {
+            // Skip the `Denomination` type as it doesn't show up in the public api
```
Comment (tcharding) [inline units/tests/api.rs:302]: This comment is stale mate :)

**Review by tcharding — DISMISSED** [review]
Mad, looks good. Just the stale comment to remove then

ACK 6202d3b1426c8fe3164aeee9bf7f7e4224899214

File: units/tests/api.rs (hunk 3 above)
Comment (shinghim) [reply units/tests/api.rs]: Removed 👍🏼

**Review by apoelstra — APPROVED** [review]
ACK 04dfe8dd45fae9b55dacfe9eb0d73ea306db14ba; successfully ran local tests

**Review by tcharding — APPROVED** [review]
ACK 04dfe8dd45fae9b55dacfe9eb0d73ea306db14ba

**Pull request merged**